### ISSUE TITLE
V3 (Memcached problems)

### ISSUE THREAD

**Issue by mikeruss1** [CIDRAM/CIDRAM#368]
I am happy with the reliability of v3, its not breaking, its not rejecting stuff it shouldnt, and as far as I can tell its not accepting stuff it shouldnt, but the latter is difficult to prove 100%. I am using it now for the home page which attracts a lot of rubbish.

the only significant issue for me is memcached, I like the performance improvement and prefer not to commit completely without it.

I have tried memcache again with the same results as last time,
in the front end tracking and stats are empty, and the cache page just has a Memcached entry with no detail, plus a link to delete it.
incidentally v2 with memcached shows "the cache is empty" on the cache page
also noticed that v3 log page with memcache only shows the first entry, thats very strange as the log isnt cached and v3 without memcache is working normally. No config change apart from memcached equals true.

the cache itself looks OK,  there is IP tracking data and "CIDRAM_Statistics-Since" contains 1669832913
thats presumably a unix date/time = I cant see any traffic count variable

**Comment by eurobank** [comment]
@mikeruss1 

On the other hand in v3, Redis seems to work a little better. Statistics work, Cache data works. 

IP tracking page doesn't work.

If you have the option for Redis, you could try it.

**Comment by mikeruss1** [comment]
My site uses memcache a lot. Would prefer not to switch really. It will likely be fixed soon.

**Comment by Maikuolan** [comment]
So.. Some limited progress on the Memcached situation (i.e., with data, which we know is being correctly stored, being not correctly recalled by the cache). Currently, the cache handler (when using Memcached) relies on the `getAllKeys` function to fetch all currently stored cache entries, which can then be processed for whatever purpose they may be needed for at the time. I've read just today, at some issues at the GitHub repository for Memcached, as well as at a few StackOverflow questions, that apparently, `getAllKeys` is broken, has been broken for a long time now, and also isn't appropriate for uses beyond diagnosis/debugging. I didn't know that until today. No mention about such things at the official documentation for the function in question at all, and the way the documentation is presented definitely implies that it's a perfectly functional function with nothing in particular wrong with it, and nothing to dissuade from using it. So I guess, in order to fix these cache problems, I'm going to need to rewrite that part of the cache handler, to either fetch all cache entries in some different way, or find some different way to achieve the same thing. Not sure currently how I'll go about that exactly, but I'll think on it a little and try to figure something out.

**Comment by eurobank** [comment]
Yeap, i searched that also and it seems getAllKeys is for debuging only, `get` family of functions should be used (so some gurus say).

**Comment by mikeruss1** [comment]
If you do getversion() before getallkeys() it should hopefully work. It works for me anyway.cheers.... Mike

**Comment by mikeruss1** [comment]
Prob more a work around than a permanent fix though. cheers.... Mike

**Comment by mikeruss1** [comment]
just a thought, there may be performance and security issues with getallkeys()

my memcache has an awful lot of entries. I dont think anything is confidential, but some folks might have a problem.

I dont think you can be selective about which keys you get?
also it raises the potential of an accidental overwrite ?

**Comment by Maikuolan** [comment]
> my memcache has an awful lot of entries. I dont think anything is confidential, but some folks might have a problem.

Strictly speaking, from a Memcached perspective, that's 100% true. Memcached (much the same as APCu in this regard) doesn't provide any kind of security or authentication in its own right: If you have access to the port number and hostname where a Memcached server is listening, you'll be able to communicate with that Memcached server, add entries, delete entries, or whatever else (which is why, if you're ever in the position of sysadmin, or able to control what's open/closed at a server, it's best to not have its ports open to the wider internet or to the network DMZ, and to ensure that it can only be accessed from within the local network, or otherwise restricted to whatever needs access to it). I'd *like* to believe that a webhost supporting it would implement appropriate safeguards of their own, such as perhaps running any particular account instance from within a virtual machine or something to give each user their own instance, thus being able to provide some privacy from each other's data, but not a good assumption to make without any actual evidence of it being the case. The people running the servers potentially mightn't know about the security risks, or, it's also quite possible that in some cases, they *do* know about the risks, and have opted to just disable Memcached at the servers in question entirely (thus maybe why it isn't available at all at some major hosting providers). Ultimately just speculating there though.

> I dont think you can be selective about which keys you get?

That's correct, too.

**Comment by mikeruss1** [comment]
I was thinking more locally. Getallkeys only retrieves keys from my domain.Maybe go for a master key of keys?

**Comment by eurobank** [comment]
Is this issue, related to the IP Tracking empty page also (when memcacheD) ?

**Comment by Maikuolan** [comment]
Yep. :-)

Each IP tracking entry is stored as its own cache entry (since v3 onward; earlier versions grouped everything together, but that could potentially lead to corruption and loss of all IP tracking data entirely when too many entries were stored; keeping separate entries helps prevent that particular problem), and because of that, it's not so easy to predict exactly what each entry would be named as, so instead of calling specific entries, the IP tracking page would call all cache entries, and then check at processing whether the cache entry called belongs to IP tracking or something else, processing IP tracking entries as necessary and ignoring unrelated data. But, because the mechanism used by the cache handler for reading all entries when using Memcached relies on a function which we now know apparently doesn't work properly, the page shows no IP tracking data, even when we know the data actually exists.

That's why I think the empty IP tracking page problem should disappear (fingers crossed) as soon as the problem with Memcached is fixed.

**Comment by Maikuolan** [comment]
Contemplating maybe implementing manual indexing for entries (basically, whenever a new entry is created, to also add the name of that entry to a separate entry, which acts as a list of all relevant entries), which could then be called explicitly in order to get a list of which other entries need to be called, which should work in theory, but could become a little messy in its own right, and then, if that indexing became corrupted or overwritten, would probably result in the same scenario, whereby the data exists, but can't be fetched anymore, and ends up with an empty page. So, not 100% sure about that idea. Might go with that if I don't end up coming up with anything better, but will try to find something better if possible.

**Comment by mikeruss1** [comment]
just a thought, you could store everything in 1 data item, serialized ? Default max size is 1 MB, should be easy to construct and retrieve further sub items if needed.

**Comment by mikeruss1** [comment]
Is the new version ok for production? I don't have memcache on my test machine. Anyone know a way of running memcache on windows?

**Comment by eurobank** [comment]
oops !!!!!


![Screenshot(1)](https://user-images.githubusercontent.com/4467158/212477344-da0b787a-bca4-48f1-9f64-84627747efc5.png)

**Comment by Maikuolan** [comment]
> Is the new version ok for production?

It *should* be. I've eyeballed over the changes several times and have been as careful as one can reasonably be when in the absence of testing, but as of yet, I still haven't been able to get Memcached working properly on my dev machine, which means I haven't been able to test the changes thoroughly, thus the "experimental bug-fix" commit name.

At the very least, the changes only target Memcached users, so anyone using anything else shouldn't notice anything different anyway. For anyone using Memcached, fingers crossed, the changes should fix all the problems we've encountered which stem from the way CIDRAM was trying to leverage Memcached, and which prevented CIDRAM from being able to display anything at the cache data page or IP tracking page.

I won't know for sure until someone actually tests it and reports back here though, unfortunately.

**Comment by Maikuolan** [comment]
> oops !!!!!

Was that from *before* or *after* the update? '^.^

**Comment by eurobank** [comment]
> > oops !!!!!
> 
> Was that from _before_ or _after_ the update? '^.^

After the update. Before it was only a "clear all".

**Comment by Maikuolan** [comment]
> After the update. Before it was only a "clear all".

That's a relief. For a brief moment there, when I saw "oops", I was thinking, *"Oh no.. Did I just fix something which didn't need to be fixed?!"*. '^.^

And, as a bonus.. If it's actually showing entries properly now, I guess that means the experimental bug-fix has worked as intended. :-)

**Comment by eurobank** [comment]
IP Tracking is still empty. And statistics also.

**Comment by Maikuolan** [comment]
> IP Tracking is still empty.

Hm.. Okay. I guess, only *partially* fixed then, in that case. At least, there's some progress. :-)

I'll continue tinkering with it to see what can be done.

**Comment by mikeruss1** [comment]
tried it - the cache page has the entries with what looks like a valid entry for a block. But IP Tracking page is empty. Stats page in cache has an entry for since. But no stats numbers. Tor working too. Incidentally my v3 non cache version is still showing just 1 for all the numbers.

**Comment by Maikuolan** [comment]
Did a bit of tinkering. Any better now, or still the same..?

**Comment by eurobank** [comment]
> Did a bit of tinkering. Any better now, or still the same..?

![Screenshot(1)](https://user-images.githubusercontent.com/4467158/212694470-72fec7da-4489-44f1-ba75-ea8b1b4e29f1.png)

**Comment by mikeruss1** [comment]
basically no. However the entry in the cache page has updated the number of infractions, so the cache itself is working. Something else odd, I looked at the log and there was only 1 entry from early this morning. Reverted to standard v3 and the log has several entries, including my recent test. The log itself isnt cached is it .. so something else is going on in retrieving data for the cache version?

**Comment by Maikuolan** [comment]
I think I've figured out now why the IP tracking entries weren't displaying at all, though there are still a few problems to figure out, like how to derive expiry values from mechanisms which don't return them at all when fetching entries. Still no idea about the statistics page problems. Will reply back again when there's more progress.

**Comment by eurobank** [comment]
> I think I've figured out now why the IP tracking entries weren't displaying at all, though there are still a few problems to figure out, like how to derive expiry values from mechanisms which don't return them at all when fetching entries. Still no idea about the statistics page problems. Will reply back again when there's more progress.

It seems to semi-work:

![Screenshot(1)](https://user-images.githubusercontent.com/4467158/212861211-d7672f5a-933b-41e4-8dd8-79e8c134aea9.png)

**Comment by mikeruss1** [comment]
confirmed. The simple log page only returns the first entry, entries per page set at 20. Cache and non cache.

**Comment by eurobank** [comment]
Also in IP Tracking:

1. Clearing works
2. MANUALLY adding some IPs, doesn't work right, shows only the FIRST auto-added.

![Screenshot(1)](https://user-images.githubusercontent.com/4467158/212870327-aefa7ee9-e889-4623-a51b-63d8aeb980eb.png)

**Comment by eurobank** [comment]
> confirmed. The simple log page only returns the first entry, entries per page set at 20. Cache and non cache.

yes only the first IP is shown.

**Comment by eurobank** [comment]
Ok, some more information.

In IP tracking details some more appeared now. Still 5 are displayed but 6 are tracked.

![Screenshot(1)](https://user-images.githubusercontent.com/4467158/212891768-d371965b-df37-432e-b1ce-07697bb2502b.png)

**Comment by eurobank** [comment]
If anyone is testing this, he MUST FIRST clear all cache in CIDRAM and all IP tracking.

After that the IP tracking seems to work better, but not 100%. Manual adding IP to tracking, doesn't show, but it IS tracked.

Expiry also is empty with no data available message.

**Comment by mikeruss1** [comment]
there are still a few problems to figure out, like how to derive expiry values from mechanisms which don't return them at all when fetching entries.
--------
been here, memcache expiry is tricky, its not visible and when you replace an item you have to provide an expiry. I came across this with rate limiting, I had to estimate the updated expiry. I guess the alternative is to also include a time stamp in the data.

**Comment by eurobank** [comment]
The IP Tracking Page with empty Expirations is a GLOBAL issue, not related to memcached. 

Because the issue is happening also with Redis BUT also with PDO !! and i guess it is not complicated to get the data from a sql Db, as with object caching memcached and redis? So it is a bug.

See the PDO:

![Screenshot(1)](https://user-images.githubusercontent.com/4467158/213149097-1bb16e2e-f1c2-4940-a31f-c8e770f103b3.png)

**Comment by Maikuolan** [comment]
Oh wow.. Didn't realise it was doing that with Redis and PDO as well. :-/

Not sure yet about how to fetch expiries from Memcached, but yeah, it definitely should be easy to do with PDO. It should be doing that already, actually, but I guess, judging from what we can see here, it isn't.

I'll investigate that tonight.

**Comment by Maikuolan** [comment]
PDO seems to be working correctly at my end.

Just in case it's an issue with drivers (e.g., depending on which drivers PDO is using, maybe it's behaving differently, returning results formatted in an unexpected way, or something like that), if you go to the homepage and click the `Extension➡Installed/Available` menu, in the list that appears, what do you see shown alongside PDO?

For example: In my case, at my dev machine, it shows "mysql, sqlite".

![Untitled](https://user-images.githubusercontent.com/12571108/213248693-2f4063a7-0fe3-446d-8b32-304cc1d70170.png)

**Comment by eurobank** [comment]
It doesn't show ANY extensions for me, while most ARE available.

![Screenshot(1)](https://user-images.githubusercontent.com/4467158/213250506-213466ed-6952-4ac9-9f3a-5ce7b0197728.png)

**Comment by eurobank** [comment]
But in Config, they all show:

![Screenshot(1)](https://user-images.githubusercontent.com/4467158/213250832-ef15aae4-fb21-4ec8-964a-b26ab53560a7.png)

**Comment by eurobank** [comment]
Scap the 2 replies i deleted.

They all show:

![Screenshot(1)](https://user-images.githubusercontent.com/4467158/213251238-d7ab7f26-763a-4640-b7eb-08b98d7b928a.png)

**Comment by Maikuolan** [comment]
Latest commit should fix the problem for Redis. Haven't yet been able to progress at all with Memcached or PDO (PDO drivers at your end look normal, and are drivers I've worked with before, so I don't think the drivers are the problem, but with that being the case, I've got no clue right now what else could be causing the problem, as looking over all the code, everything *seems* good to me, and it's still working properly at my end, so.. I'll keep digging to get to the bottom of it, but no idea currently; as for Memcached, I've gone digging right through their entire API, and can't find anything in their API at all for returning the expiries for entries, so I guess, if I want to fix the problem with Memcached, I'll need to either keep shadow records for expiries, or build some kind of parallel system of my own for it otherwise; will figure something out anyway).

**Comment by mikeruss1** [comment]
there is no facility to access the expiry time in Memcached - If you think the problem is more trouble than its worth, I can switch to Redis ?

**Comment by eurobank** [comment]
removed

**Comment by mikeruss1** [comment]
setup my wampserver home development setup to use redis. Updated to the latest CIDRAM version. Tracking is working and stats is working. Not an extensive test but basically it looks ok. 

Info via Google if anyone wants to try it. I sandboxed the redis server just in case!.
I dont think there is a way to do this with memcached.

**Comment by eurobank** [comment]
> setup my wampserver home development setup to use redis. Updated to the latest CIDRAM version. Tracking is working and stats is working. Not an extensive test but basically it looks ok.
> 

What vs of php is that ?

**Comment by mikeruss1** [comment]
8.1.8

**Comment by eurobank** [comment]
> 8.1.8

i wonder if this is not working for me because of php 7.4.x .....

**Comment by mikeruss1** [comment]
As long as you have the correct redis DLL for 7.4 you should be ok I think

**Comment by eurobank** [comment]
Actually after switching to 8.0.x it seems to work:

`System information:

    CIDRAM version used: 3.0.0
        Branch latest stable: 3.0.0-beta2 – Latest stable: 2.9.2 – Latest unstable: 3.0.0
    PHP version used: 8.0.25
        Branch latest stable: 8.0.27 – Latest stable: 8.2.1 – Latest unstable: 8.3.0
        SAPI used: litespeed
    Extension➡Installed/Available
        PCRE➡Yes (8.0.25)
        cURL➡Yes (8.0.25)
        APCu➡No
        Memcached➡Yes (3.1.5)
        Redis➡Yes (5.3.3)
        PDO➡Yes (8.0.25), {mysql, odbc, sqlite}
`

![Screenshot(3)](https://user-images.githubusercontent.com/4467158/213863349-aae5bf74-147e-4545-b853-f9e214a7792e.png)

**Comment by mikeruss1** [comment]
have put it into prod using php 7.4.33 and its working. There has been a new version of common classes which may well have fixed it.

**Comment by eurobank** [comment]
> have put it into prod using php 7.4.33 and its working. There has been a new version of common classes which may well have fixed it.

This is strange. Could my issue be related to the fact that the CIDRAM is outside of ANY website and that means that i can't select PHP version for non-websites? 

I mean CIDRAM shows the version of the SITE the loader call is in. BUT the CIDRAM files is unclear what version they run. Not sure if you get what i try to say.

@Maikuolan what do you think?

**Comment by eurobank** [comment]
> have put it into prod using php 7.4.33 and its working. There has been a new version of common classes which may well have fixed it.

I doubt my issue is a "driver" or common classes or whatever update. Since i;m under CloudLinux and all updates are automatically installed as soon as they become available. And NEVER had any issues with all that from day one i used the hosting services (8 years now) i'm on.

**Comment by mikeruss1** [comment]
Common classes is a component of cidram.

**Comment by eurobank** [comment]
I moved CIDRAM 3 into a website path just to rule out any issues with php versions. The issue is still there.

So it is something else.

**Comment by Maikuolan** [comment]
Sorry for the delayed reply. Busy with other stuff yesterday.

I can't guarantee any extensions working as expected with PHP 8.2 just yet, because Windows-compatible binaries for those extensions aren't available for PHP 8.2 just yet (I'll be able to test them all once they become available).

They *should* all work properly for every other supported PHP branch though (i.e., 7.2, 7.3, 7.4, 8.0, 8.1).

> Common classes is a component of cidram.

Needed for CIDRAM to work properly, too. Main reason I've listed it as its own component, rather than just bundling the relevant files along with the rest of CIDRAM's core files, is for easier maintainability, because the specific classes included with the "common classes package", I'm also using with a few other projects I wrote and maintain (e.g., phpMussel), so, given they exist across more than one project, it makes sense to distinguish them from the other core files.

> there is no facility to access the expiry time in Memcached - If you think the problem is more trouble than its worth, I can switch to Redis ?

As much as I'm not a fan of giving up, I may need to give up on the Memcached side of things, at least, given there's no reliable way to extract TTLs from Memcached entries, and providing shadow entries for the entire cache is going to bloat it all like crazy, and though I *could* go that approach if I *had* to for the sake of being able to properly display expiries at the IP tracking page, I'm not so sure it's worthwhile, giving that caching should still generally work all the same in any case, and it's just the ability to properly display expiries affected there.

I'd definitely still like to be able to get to the bottom of why the other caching mechanisms aren't working at @eurobank's installation, given that the problem of lacking a way to extract TTLs doesn't exist for the other caching mechanisms, and those other caching mechanisms seem to work correctly at my end. So, there's definitely something here I'm missing, that I'm just not seeing. If I can figure out what it is, I imagine it should be possible to resolve the problem.

> This is strange. Could my issue be related to the fact that the CIDRAM is outside of ANY website and that means that i can't select PHP version for non-websites?

In theory, that shouldn't be a problem at all (but given we aren't absolutely sure what's going on here, I wouldn't 100% rule anything out just yet). The reason I say that it shouldn't be a problem at all, is that, although CIDRAM is stored outside any website, the actual entrypoints are a different story, and with the way that PHP handles "included" code (i.e., when you have one PHP file, which "includes in" or "requires in" code from some different PHP file), PHP treats the entire code stack as if originating from the first file in the stack (i.e., the entrypoint). That's why, as long as a PHP file is "included" or "required" from some other PHP file stored within a directory with the necessary permissions to execute PHP code, you can still "execute" that "included" or "required" code, even if the file where it comes from is stored within a directory without the necessary permissions.

In other words.. In theory, at least, the PHP version "selected" for CIDRAM should be the same as whichever PHP version is selected for the website where the entrypoint exists.

**Comment by Maikuolan** [comment]
> I moved CIDRAM 3 into a website path just to rule out any issues with php versions. The issue is still there.
> 
> So it is something else.

Good to be able to rule it out. :-)



> I doubt my issue is a "driver" or common classes or whatever update. Since i;m under CloudLinux and all updates are automatically installed as soon as they become available. And NEVER had any issues with all that from day one i used the hosting services (8 years now) i'm on.

I don't think it's drivers, either. As long as the currently installed version of the common classes package is the current latest available version, it shouldn't be that either, seeing as that's the version I'm working from and I'm not encountering the same problem.

**Comment by eurobank** [comment]
> In other words.. In theory, at least, the PHP version "selected" for CIDRAM should be the same as whichever PHP version is selected for the website where the entrypoint exists.

I tested that, move it inside a website, same issue. So ..

Let me know if i can help using some code/snippet to resolve this. I'm willing to test ON my system. You have my contact details, free free to use them (or not).

cheers

**Comment by eurobank** [comment]
> 
> > This is strange. Could my issue be related to the fact that the CIDRAM is outside of ANY website and that means that i can't select PHP version for non-websites?
> 
> In theory, that shouldn't be a problem at all (but given we aren't absolutely sure what's going on here, I wouldn't 100% rule anything out just yet). The reason I say that it shouldn't be a problem at all, is that, although CIDRAM is stored outside any website, the actual entrypoints are a different story, and with the way that PHP handles "included" code (i.e., when you have one PHP file, which "includes in" or "requires in" code from some different PHP file), PHP treats the entire code stack as if originating from the first file in the stack (i.e., the entrypoint). That's why, as long as a PHP file is "included" or "required" from some other PHP file stored within a directory with the necessary permissions to execute PHP code, you can still "execute" that "included" or "required" code, even if the file where it comes from is stored within a directory without the necessary permissions.
> 
> In other words.. In theory, at least, the PHP version "selected" for CIDRAM should be the same as whichever PHP version is selected for the website where the entrypoint exists.

Maybe there is some issue in there.

Please let me know what you think:

As you know i monitor 5 sites with ONE instance of CIDRAM. 2 Wordpress sites run PHP 8.0.x and 3 run PHP 7.4.x for compatibility reasons. 

So CIDRAM is "called" (prepended) from DIFFERENT PHP version sites. Could that be an issue?

**Comment by mikeruss1** [comment]
**"I may need to give up on Memcached"**
-----------
I have switched some traffic to v3 and redis and it looks OK so far, give it a few days and if all OK I will move everything across. So all being well I wont need memcached.

**Comment by eurobank** [comment]
Some update. I have switched all sites of mine to php 7.4x. and the issues remained. As they were.

So i'm ruling out any compatibility or setup issues.

**Comment by eurobank** [comment]
> ## **"I may need to give up on Memcached"**
> 
> I have switched some traffic to v3 and redis and it looks OK so far, give it a few days and if all OK I will move everything across. So all being well I wont need memcached.

yeah, it seems Redis also works 100% for me, no matter what. So for now, i guess, i will keep Redis from all options.

**Comment by Maikuolan** [comment]
Fixed for PDO now, too (#383, #390). So, the problem now persists only for Memcached.

There's some successful fixes attached to this issue (e.g., ealier, with Redis), but I think I should still add the "Won't Fix" label, due to that the core of the issue (i.e., Memcached not displaying expiries properly) hasn't been fixed, and looks to be somewhat unfixable (beyond hacky workarounds like shadow caching, that is).

I don't think there was anything else, so I'll go ahead and close this now. If I've missed something, let me know, and we can reopen it.
